include/opae/hash_map.h

```diff
@@ -0,0 +1,118 @@
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#ifndef __OPAE_HASH_MAP_H__
+#define __OPAE_HASH_MAP_H__
+#include <stdint.h>
+#include <stdbool.h>
+#include <opae/types_enum.h>
+
+#ifndef UNUSED_PARAM
+#define UNUSED_PARAM(x) ((void)(x))
+#endif // UNUSED_PARAM
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
+typedef enum _opae_hash_map_flags {
+	OPAE_HASH_MAP_UNIQUE_KEYSPACE = (1u << 0)
+} opae_hash_map_flags;
+
+typedef struct _opae_hash_map_item {
+	void *key;
+	void *value;
+	struct _opae_hash_map_item *next;
+} opae_hash_map_item;
+
+typedef struct _opae_hash_map {
+	uint32_t num_buckets;
+	uint32_t hash_seed;
+	opae_hash_map_item **buckets;
+	int flags;
+	void *cleanup_context;
+	uint32_t (*key_hash)(uint32_t num_buckets,	   ///< required
+			     uint32_t hash_seed,
+			     void *key);
+	int (*key_compare)(void *keya, void *keyb);	   ///< required
+	void (*key_cleanup)(void *key, void *context);	   ///< optional
+	void (*value_cleanup)(void *value, void *context); ///< optional
+} opae_hash_map;
+
+fpga_result opae_hash_map_init(opae_hash_map *hm,
+			       uint32_t num_buckets,
+			       uint32_t hash_seed,
+			       int flags,
+			       uint32_t (*key_hash)(uint32_t num_buckets,
+						    uint32_t hash_seed,
+						    void *key),
+			       int (*key_compare)(void *keya, void *keyb),
+			       void (*key_cleanup)(void *key, void *context),
+			       void (*value_cleanup)(void *value, void *context));
+
+fpga_result opae_hash_map_add(opae_hash_map *hm,
+			      void *key,
+			      void *value);
+
+fpga_result opae_hash_map_find(opae_hash_map *hm,
+			       void *key,
+			       void **value);
+
+fpga_result opae_hash_map_remove(opae_hash_map *hm,
+				 void *key);
+
+fpga_result opae_hash_map_destroy(opae_hash_map *hm);
+
+bool opae_hash_map_is_empty(opae_hash_map *hm);
+
+static inline
+uint32_t opae_u64_key_hash(uint32_t num_buckets,
+			   uint32_t hash_seed,
+			   void *key)
+{
+	UNUSED_PARAM(hash_seed);
+	uint64_t ukey = (uint64_t)key;
+	return (uint32_t)(ukey % num_buckets);
+}
+
+static inline
```

Comment:
I don't know what the compiler does when it sees this in the header file as a static. Won't you get a separate instance of opae_u64_key_compare in each module, so that the vfio code will pass in the address of it's private copy? Then the comparison in hash_map.c will compare against it's private copy and won't match, which will cause the test to see whether to invoke opae_u64_key_compare fail and it will call through the pointer.